Augment the content rule list allowlist mechanism introduced in 259930@main

whsieh opened the pull request:
#### f9309fb655519c8b3aa0c2ea92813f5c97c44ee7
<pre>
Augment the content rule list allowlist mechanism introduced in 259930@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=251922">https://bugs.webkit.org/show_bug.cgi?id=251922</a>
rdar://104453682

Reviewed by Alex Christensen.

The previous change in <a href="https://commits.webkit.org/259930@main">https://commits.webkit.org/259930@main</a> added a way to configure content rule
list enablement in the following three ways, as a per-navigation switch:

1. Unconditionally enable all content rule lists.
2. Unconditionally disable all content rule lists.
3. Enable a specific subset of content rule lists (by identifier).

Unfortunately, this turned out to be insufficient to fulfill the internal client&apos;s use case, which
requires a fourth enablement configuration that&apos;s not (easily) expressible through existing means:

4. Disable a specific subset of content rule lists (by identifier).

To make it possible to configure all 4 of the above on a per-navigation basis, we replace both the
existing `-_contentBlockersEnabled` flag and the newer `-_disabledContentRuleListIdentifiers`
property with a single SPI:

```
- (void)_setContentRuleListsEnabled:(BOOL)enabled exceptions:(NSSet&lt;NSString *&gt; *)exceptions;
```

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::contentExtensionEnablement const):
(WebCore::DocumentLoader::setContentExtensionEnablement):
(WebCore::DocumentLoader::disabledContentExtensions const): Deleted.
(WebCore::DocumentLoader::setDisabledContentExtensions): Deleted.

Rename `DisabledContentExtensions` to `ContentExtensionEnablement`, and change it from a variant to
a pair (now that *both* the enum type and `HashSet` are required in order to express all potential
configurations).

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reload):
* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::ruleListFilter):
(WebCore::UserContentProvider::processContentRuleListsForLoad):
(WebCore::contentRuleListsEnabled): Deleted.
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::encode const):
(WebKit::WebsitePoliciesData::decode):

More renaming: `WebCore::DisabledContentExtensions` -&gt; `WebCore::ContentExtensionEnablement`.

(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.cpp:
(API::WebsitePolicies::copy const):
(API::WebsitePolicies::data):
(API::WebsitePolicies::contentBlockersEnabled const): Deleted.
(API::WebsitePolicies::setContentBlockersEnabled): Deleted.
(API::WebsitePolicies::setDisabledContentRuleListIdentifiers): Deleted.
(API::WebsitePolicies::disabledContentRuleListIdentifiers const): Deleted.
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/C/WKWebsitePolicies.cpp:
(WKWebsitePoliciesSetContentBlockersEnabled):
(WKWebsitePoliciesGetContentBlockersEnabled):
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setContentBlockersEnabled:]):
(-[WKWebpagePreferences _contentBlockersEnabled]):

Reimplement these in terms of the new SPI, by treating these as default-enablement (or default-
disablement) with no rule list exceptions.

(-[WKWebpagePreferences _setContentRuleListsEnabled:exceptions:]):
(-[WKWebpagePreferences _setDisabledContentRuleListIdentifiers:]): Deleted.
(-[WKWebpagePreferences _disabledContentRuleListIdentifiers]): Deleted.

Replace the property added in 259930@main with a set-only method instead.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:

Adjust an existing API test to check for both default-enablement (with an exception rule list ID to
disable), as well as default-disablement (with an exception rule list ID to enable).

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView synchronouslyLoadRequest:preferences:]):

Add a new API test helper method on `TestWKWebView`.

Canonical link: <a href="https://commits.webkit.org/260084@main">https://commits.webkit.org/260084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c78683483e66860ea8288615e0a19888a70dc341

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106954 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39747 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116133 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115673 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110855 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17472 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7187 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99139 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112719 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13223 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96250 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40821 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95115 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27892 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9133 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29244 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6243 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15300 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48803 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6964 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11242 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/UserContentProvider.cpp:
```diff
@@ -112,30 +112,43 @@ static ContentExtensions::ContentExtensionsBackend::RuleListFilter ruleListFilte
+    auto& exceptions = mainLoader->contentExtensionEnablement().second;
+    switch (mainLoader->contentExtensionEnablement().first) {
+    case ContentExtensionDefaultEnablement::Disabled:
+        if (exceptions.isEmpty()) {
```
This is probably just adding unnecessary complexity.  HashSet.contains is pretty fast when the set is empty.

whsieh replied:
Sounds good — simplified this logic.

achristensen07 commented on Source/WebCore/page/UserContentProvider.cpp:
```diff
@@ -112,30 +112,43 @@ static ContentExtensions::ContentExtensionsBackend::RuleListFilter ruleListFilte
 }
 
-static bool contentRuleListsEnabled(DocumentLoader& documentLoader)
+static bool allContentRuleListsAreDisabled(DocumentLoader& documentLoader)
```
This check should probably just be removed.  Adds unneeded complexity.

whsieh replied:
Good point! Removed.

whsieh committed:
Augment the content rule list allowlist mechanism introduced in 259930@main
https://bugs.webkit.org/show_bug.cgi?id=251922
rdar://104453682

Reviewed by Alex Christensen.

The previous change in https://commits.webkit.org/259930@main added a way to configure content rule
list enablement in the following three ways, as a per-navigation switch:

1. Unconditionally enable all content rule lists.
2. Unconditionally disable all content rule lists.
3. Enable a specific subset of content rule lists (by identifier).

Unfortunately, this turned out to be insufficient to fulfill the internal client's use case, which
requires a fourth enablement configuration that's not (easily) expressible through existing means:

4. Disable a specific subset of content rule lists (by identifier).

To make it possible to configure all 4 of the above on a per-navigation basis, we replace both the
existing `-_contentBlockersEnabled` flag and the newer `-_disabledContentRuleListIdentifiers`
property with a single SPI:

```
- (void)_setContentRuleListsEnabled:(BOOL)enabled exceptions:(NSSet<NSString *> *)exceptions;
```

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::contentExtensionEnablement const):
(WebCore::DocumentLoader::setContentExtensionEnablement):
(WebCore::DocumentLoader::disabledContentExtensions const): Deleted.
(WebCore::DocumentLoader::setDisabledContentExtensions): Deleted.

Rename `DisabledContentExtensions` to `ContentExtensionEnablement`, and change it from a variant to
a pair (now that *both* the enum type and `HashSet` are required in order to express all potential
configurations).

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reload):
* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::ruleListFilter):
(WebCore::UserContentProvider::processContentRuleListsForLoad):
(WebCore::contentRuleListsEnabled): Deleted.
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::encode const):
(WebKit::WebsitePoliciesData::decode):

More renaming: `WebCore::DisabledContentExtensions` -> `WebCore::ContentExtensionEnablement`.

(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.cpp:
(API::WebsitePolicies::copy const):
(API::WebsitePolicies::data):
(API::WebsitePolicies::contentBlockersEnabled const): Deleted.
(API::WebsitePolicies::setContentBlockersEnabled): Deleted.
(API::WebsitePolicies::setDisabledContentRuleListIdentifiers): Deleted.
(API::WebsitePolicies::disabledContentRuleListIdentifiers const): Deleted.
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/C/WKWebsitePolicies.cpp:
(WKWebsitePoliciesSetContentBlockersEnabled):
(WKWebsitePoliciesGetContentBlockersEnabled):
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setContentBlockersEnabled:]):
(-[WKWebpagePreferences _contentBlockersEnabled]):

Reimplement these in terms of the new SPI, by treating these as default-enablement (or default-
disablement) with no rule list exceptions.

(-[WKWebpagePreferences _setContentRuleListsEnabled:exceptions:]):
(-[WKWebpagePreferences _setDisabledContentRuleListIdentifiers:]): Deleted.
(-[WKWebpagePreferences _disabledContentRuleListIdentifiers]): Deleted.

Replace the property added in 259930@main with a set-only method instead.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:

Adjust an existing API test to check for both default-enablement (with an exception rule list ID to
disable), as well as default-disablement (with an exception rule list ID to enable).

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView synchronouslyLoadRequest:preferences:]):

Add a new API test helper method on `TestWKWebView`.

Canonical link: https://commits.webkit.org/260084@main
